coreax/approximation.py

```diff
@@ -180,7 +183,7 @@ def gramian_row_mean(self, x: ArrayLike, **kwargs) -> Array:
         :return: Approximation of the base kernel's Gramian row-mean
         """
         del kwargs
-        data = jnp.atleast_2d(x)
+        data = jnp.asarray(x)
```

Comment:
In all previous cases, we were assuming data was handled as a 2D JAX array (with `jnp.atleast_2d`). I assume `jnp.asarray` is needed for compatibility with `x` if `x` is of type `Data`? Can we not keep the assumption that everything is worked with as a 2D array?

Reply:
We only test for 2D at the moment. I don't like the usage of `atleast_2d` as it is not explicit to the user where this dimension will be added (the leading axis). If the user passes an n-vector (believing this represents n scalar observations) then `atleast_2d` will silently give them the wrong result. This is because `atleast_2d` converts the vector to an array with shape `(1, n)`, rather than `(n, 1)`. 

The difficulty is that different users may assume different behaviour for vectors. Some may (correctly in this case) assume an n-vector represents a single n-dimensional observation, whereas others may (incorrectly in this case) assume an n-vector represents n 1-dimensional observations.

Because of this, I think we only have three options: 
1. throw an error when passed an array that is not 2D `if jnp.ndim(x) != 2: raise ValueError("Must be 2D")`
2. allow k-dimensional arrays and let JAX handle any resulting incompatible mathematical operations.
3. use beartype and jaxtyping to provide (optional) runtime type checking of the array dimensions.

If we start using jaxtyping everywhere then the expectations on user data become clearer before runtime (at programming time).

Addressing this issue properly across the entire codebase is out of scope for this pull, so I would prefer to keep this as is, or just wrap the `asarrays` in an `atleast_2d` just to preserve the previous behaviour (whichever you would prefer).

Reply:
@tc85324 Thanks for highlighting the options. I agree that it is not explicit to the user what will happen with particular inputs and could lead to confusion. My proposal is to wrap the `asarrays` in an `atleast_2d` call to preserve the previous behavior for now, but I will also create a ticket that details this discussion and in resolving that, we do one of the three listed options and remove the wrapped `atleast_2d` calls. How does that sound?

Reply:
For reference this is now discussed in issue #630